bindings/python/slimt.cpp

```diff
@@ -142,6 +144,103 @@ class PyService {
+    // # Case 1: Byte Ranges
+    //
+    // source = response.source.text // "Hello world"
+    // range = response.source.word_as_range(0, 0)
+    //
+    // #  Additional encoding step required to access
+    // #  "byte" ranges.
+    // bytes = source.encode("utf8")
+    //
+    // word_bytes = bytes[range.begin:range.end]
+    // word = word_bytes.decode("utf8")
+    //
+    // # Case 2: Unicode (UTF-8 ranges)
+    //
+    // What we want to do, is omit the conversion step, by providing ranges
+    // native to python.
+    //
+    // source = response.source.text // "Hello world"
+    // range = response.source.word_as_range(0, 0)
+    // word = source[range.begin:range.end]
+    //
+    //
+    // Assert:
+    //  word == response.source.word(0, 0)
+    //  [python-slicing] = C++ string conversion
+    //
+    // Why? Diagnosis?
+    //
+    // pybind11 converts std::string, by default to unicode-strings (python).
+    // However, since we manage ranges, they remain "byte ranges". We can
+    // mitigate this problem, by patching the annotation's ranges to point to
+    // unicode before returning the response.
+
+    auto transform_to_utf8 =
+        [](const AnnotatedText &annotated) -> std::vector<Range> {
+      // f: Range [Bytes] -> Range[UTF8]
+      const std::string &text = annotated.text;
+
+      size_t sentence_idx = 0;
+      size_t word_idx = 0;
+      Range current = annotated.word_as_range(sentence_idx, word_idx);
+
+      std::vector<Range> words;
+
+      size_t utf8_idx = 0;
+      size_t byte_idx = 0;
+      Range utf8{
+          //
+          .begin = utf8_idx,  //
+          .end = 0            //
+      };
+      // This loops run on the entire string.
+      //
+      // We have indices into two views of the same string. One is bytes, the
+      // other is utf8 encoded. We want to convert what is bytes to utf8.
+      //
+      // For this, we traverse through the characters, checking for unicode
+      // encoding and associated adjustments to the utf8_idx, keeping
+      // correspondences with the byte_idx;
+      for (const char c : text) {
+        // current = [begin, end)
+        if ((c & 0xc0) != 0x80)  // if is not utf-8 continuation character
+          ++utf8_idx;
+
+        ++byte_idx;
+
+        if (byte_idx == current.end) {
+          utf8.end = utf8_idx;
+
+          // Push it to the list of (utf8) words.
+          // We will use these to create the utf8 range annotation.
+          words.push_back(utf8);
+
+          ++word_idx;
+          if (word_idx >= annotated.word_count(sentence_idx)) {
+            ++sentence_idx;
+            word_idx = 0;
+          }
+
+          current = annotated.word_as_range(sentence_idx, word_idx);
+
+          utf8.begin = utf8_idx;
+          utf8.end = -1;
+        }
+      }
+      return words;
+    };
+
+    response.source.update_annotation(transform_to_utf8(response.source));
+    response.target.update_annotation(transform_to_utf8(response.target));
```

Comment:
If we take a closer look, this does not require anything other than the information with `Annotation` to operate. This can therefore be a member (and it better of being so). This has the benefits of disentangling caller code, making it cleaner. We can have caller syntax as one of the following:

```cpp
AnnotatedText::to_utf8_in_place() // okay compromise

// A more over-engineered version. When we get to WebAssembly
// UTF-16 will be useful, because JavaScript uses UTF16. 
enum class Encoding {
   Byte,
   UTF8,
   UTF16
};

// AnnotatedText::transform_in_place(Encoding encoding){ 
AnnotatedText::to(Encoding encoding){
    // Conversion. 
    // There'll be N^2, but we can make it N by going through 
    // a sane intermediate.
}
```

Reply:
Just so it's clear, we're looking for a calling code:

```cpp
  static void change_ranges_to_utf8(Response &response) {
    response.source.to(Encoding::UTF8);
    response.target.to(Encoding::UTF8);
  }
```